File: platform-lib-go/registration.go
package platformlib

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"strings"
	"sync"
	"time"

	"github.com/nats-io/nats.go"
)

const (
	// Bare paths — direct S2S calls bypass the gateway; platform-api serves
	// internal endpoints at root (gateway-strip routing).
	registerPath     = "/internal/environment/register"
	snapshotPath     = "/internal/environment/snapshot"
	reRegisterPeriod = 5 * time.Minute
)

// RegistrationConfig declares this service's identity for platform-api.
type RegistrationConfig struct {
	ServiceCode   string
	SelfBaseURL   string
	PlatformURL   string
	ServiceSecret string
	SpaceTag      string
	Version       string
	ExtraPaths    []string
	Features      []string
}

// routePrefix is gateway-facing: the prefix spe-api matches before stripping.
func (c RegistrationConfig) routePrefix() string { return "/api/" + c.ServiceCode + "/**" }

// healthURL is bare — health checks hit the instance directly, which serves
// actuator at root (gateway-strip routing).
func (c RegistrationConfig) healthURL() string {
	return strings.TrimRight(c.SelfBaseURL, "/") + "/actuator/health"
}

func (c RegistrationConfig) request() RegisterRequest {
	extra := c.ExtraPaths
	if extra == nil {
		extra = []string{}
	}
	features := c.Features
	if features == nil {
		features = []string{}
	}
	return RegisterRequest{
		ServiceCode: c.ServiceCode,
		BaseURL:     c.SelfBaseURL,
		HealthURL:   c.healthURL(),
		RoutePrefix: c.routePrefix(),
		ExtraPaths:  extra,
		Version:     c.Version,
		SpaceTag:    c.SpaceTag,
		Features:    features,
	}
}

// Registrar self-registers with platform-api and keeps the local registry
// fresh. Port of PlatformRegistrationClient.
type Registrar struct {
	cfg      RegistrationConfig
	registry *LocalServiceRegistry
	http     *http.Client

	mu         sync.Mutex
	instanceID string
}

func NewRegistrar(cfg RegistrationConfig, registry *LocalServiceRegistry) *Registrar {
	return &Registrar{
		cfg:        cfg,
		registry:   registry,
		http:       &http.Client{Timeout: 10 * time.Second},
		instanceID: InstanceID(cfg.SelfBaseURL),
	}
}

func (r *Registrar) InstanceID() string {
	r.mu.Lock()
	defer r.mu.Unlock()
	return r.instanceID
}

// RegisterOnce POSTs registration then pulls the snapshot.
func (r *Registrar) RegisterOnce(ctx context.Context) error {
	body, _ := json.Marshal(r.cfg.request())
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, r.cfg.PlatformURL+registerPath, bytes.NewReader(body))
	if err != nil {
		return err
	}
	req.Header.Set("X-Service-Secret", r.cfg.ServiceSecret)
	req.Header.Set("Content-Type", "application/json")
	resp, err := r.http.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 400 {
		return fmt.Errorf("registration status %d", resp.StatusCode)
	}
	var parsed RegisterResponse
	if json.NewDecoder(resp.Body).Decode(&parsed) == nil && parsed.InstanceID != "" {
		r.mu.Lock()
		r.instanceID = parsed.InstanceID
		r.mu.Unlock()
	}
	r.PullSnapshot(ctx)
	return nil
}

func (r *Registrar) PullSnapshot(ctx context.Context) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, r.cfg.PlatformURL+snapshotPath, nil)
	if err != nil {
		return
	}
	req.Header.Set("X-Service-Secret", r.cfg.ServiceSecret)
	resp, err := r.http.Do(req)
	if err != nil {
		log.Printf("snapshot pull failed: %v", err)
		return
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 400 {
		log.Printf("snapshot pull status %d", resp.StatusCode)
		return
	}
	var snap RegistrySnapshot
	if err := json.NewDecoder(resp.Body).Decode(&snap); err != nil {
		log.Printf("snapshot decode failed: %v", err)
		return
	}
	r.registry.UpdateFromSnapshot(snap)
}

func (r *Registrar) Deregister(ctx context.Context) {
	url := fmt.Sprintf("%s%s/%s/instances/%s", r.cfg.PlatformURL, registerPath, r.cfg.ServiceCode, r.InstanceID())
	req, err := http.NewRequestWithContext(ctx, http.MethodDelete, url, nil)
	if err != nil {
		return
	}
	req.Header.Set("X-Service-Secret", r.cfg.ServiceSecret)
	if resp, err := r.http.Do(req); err == nil {
		resp.Body.Close()
		log.Printf("deregistered instance %s from platform-api", r.InstanceID())
	}
}

// RegisterWithBackoff retries the initial registration until it succeeds or
// ctx is cancelled.
func (r *Registrar) RegisterWithBackoff(ctx context.Context) {
	backoffs := []time.Duration{500, 1000, 2000, 4000, 8000, 15000, 30000}
	i := 0
	for {
		if err := r.RegisterOnce(ctx); err == nil {
			log.Printf("registered with platform-api as instance %s (tag: %s)", r.InstanceID(), tagLabel(r.cfg.SpaceTag))
			return
		} else {
			wait := backoffs[min(i, len(backoffs)-1)] * time.Millisecond
			log.Printf("platform-api registration failed: %v; retry in %v", err, wait)
			select {
			case <-time.After(wait):
				i++
			case <-ctx.Done():
				return
			}
		}
	}
}

// Run launches the background lifecycle: NATS subscriptions + periodic
// re-register. Call after RegisterWithBackoff.
func (r *Registrar) Run(ctx context.Context, bus *Bus) {
	if bus != nil {
		_, _ = bus.Subscribe("env.global.PLATFORM_RESTARTED", func(_ *nats.Msg) {
			go func() {
				time.Sleep(2 * time.Second)
				// platform-api's registry version reset on restart; drop our
				// baseline so the fresh snapshot is accepted, not rejected as stale.
				r.registry.ResetVersion()
				if r.RegisterOnce(ctx) == nil {
					log.Printf("re-registered after PLATFORM_RESTARTED")
				}
			}()
		})
		_, _ = bus.Subscribe("env.global.ENVIRONMENT_CHANGED", func(_ *nats.Msg) {
			go r.PullSnapshot(ctx)
		})
	}

	go func() {
		ticker := time.NewTicker(reRegisterPeriod)
		defer ticker.Stop()
		for {
			select {
			case <-ticker.C:
				_ = r.RegisterOnce(ctx)
			case <-ctx.Done():
				return
			}
		}
	}()
}

func tagLabel(tag string) string {
	if strings.TrimSpace(tag) == "" {
		return "untagged"
	}
	return tag
}

File: platform-lib-rs/src/registration.rs
//! Self-registration with platform-api — port of
//! `com.plm.platform.environment.PlatformRegistrationClient`.
//!
//! On start: POST a [`RegisterRequest`] to
//! `/api/platform/internal/environment/register` (retry with backoff), pull
//! the registry snapshot to bootstrap [`LocalServiceRegistry`], then keep it
//! fresh via a 5-minute periodic re-register plus NATS events
//! (`PLATFORM_RESTARTED` → re-register, `ENVIRONMENT_CHANGED` → pull). On
//! shutdown: best-effort DELETE by instance id.

use crate::dto::{RegisterRequest, RegisterResponse, RegistrySnapshot};
use crate::error::{PlatformError, Result};
use crate::nats::NatsBus;
use crate::registry::LocalServiceRegistry;
use crate::util::instance_id;
use std::sync::{Arc, Mutex};
use std::time::Duration;

// Bare paths — direct S2S calls bypass the gateway; platform-api serves
// internal endpoints at root (gateway-strip routing).
const REGISTER_PATH: &str = "/internal/environment/register";
const SNAPSHOT_PATH: &str = "/internal/environment/snapshot";
const RE_REGISTER_PERIOD: Duration = Duration::from_secs(300);

#[derive(Debug, Clone)]
pub struct RegistrationConfig {
    pub service_code: String,
    pub self_base_url: String,
    pub platform_url: String,
    pub service_secret: String,
    pub space_tag: String,
    pub version: String,
    pub extra_paths: Vec<String>,
    pub features: Vec<String>,
}

impl RegistrationConfig {
    fn route_prefix(&self) -> String {
        // Gateway-facing: the prefix spe-api matches before stripping.
        format!("/api/{}/**", self.service_code)
    }
    fn health_url(&self) -> String {
        // Bare — health checks hit the instance directly, which serves
        // actuator at root (gateway-strip routing).
        format!("{}/actuator/health", self.self_base_url.trim_end_matches('/'))
    }
    fn to_request(&self) -> RegisterRequest {
        RegisterRequest {
            service_code: self.service_code.clone(),
            base_url: self.self_base_url.clone(),
            health_url: self.health_url(),
            route_prefix: self.route_prefix(),
            extra_paths: self.extra_paths.clone(),
            version: self.version.clone(),
            space_tag: self.space_tag.clone(),
            features: self.features.clone(),
        }
    }
}

pub struct Registrar {
    http: reqwest::Client,
    registry: Arc<LocalServiceRegistry>,
    config: RegistrationConfig,
    instance_id: Mutex<String>,
}

impl Registrar {
    pub fn new(config: RegistrationConfig, registry: Arc<LocalServiceRegistry>) -> Arc<Self> {
        // Deterministic fallback id (server echoes the same value).
        let id = instance_id(&config.self_base_url);
        Arc::new(Self {
            http: reqwest::Client::new(),
            registry,
            config,
            instance_id: Mutex::new(id),
        })
    }

    pub fn instance_id(&self) -> String {
        self.instance_id.lock().unwrap().clone()
    }

    /// POST registration, then pull the snapshot. Returns Ok only if the POST
    /// succeeds.
    pub async fn register_once(&self) -> Result<()> {
        let url = format!("{}{}", self.config.platform_url, REGISTER_PATH);
        let resp = self
            .http
            .post(&url)
            .header("X-Service-Secret", &self.config.service_secret)
            .json(&self.config.to_request())
            .send()
            .await?;
        let status = resp.status().as_u16();
        if status >= 400 {
            let body = resp.text().await.unwrap_or_default();
            return Err(PlatformError::UpstreamStatus { status, body });
        }
        if let Ok(parsed) = resp.json::<RegisterResponse>().await {
            if let Some(id) = parsed.instance_id {
                *self.instance_id.lock().unwrap() = id;
            }
        }
        self.pull_snapshot().await;
        Ok(())
    }

    pub async fn pull_snapshot(&self) {
        let url = format!("{}{}", self.config.platform_url, SNAPSHOT_PATH);
        match self
            .http
            .get(&url)
            .header("X-Service-Secret", &self.config.service_secret)
            .send()
            .await
        {
            Ok(resp) if resp.status().is_success() => {
                match resp.json::<RegistrySnapshot>().await {
                    Ok(snapshot) => {
                        let v = snapshot.version;
                        let n = snapshot.services.len();
                        self.registry.update_from_snapshot(snapshot);
                        tracing::debug!("registry refreshed: {n} services (snapshot v{v})");
                    }
                    Err(e) => tracing::warn!("snapshot decode failed: {e}"),
                }
            }
            Ok(resp) => tracing::warn!("snapshot pull status {}", resp.status()),
            Err(e) => tracing::warn!("snapshot pull failed: {e}"),
        }
    }

    pub async fn deregister(&self) {
        let id = self.instance_id();
        let url = format!(
            "{}{}/{}/instances/{}",
            self.config.platform_url, REGISTER_PATH, self.config.service_code, id
        );
        let _ = self
            .http
            .delete(&url)
            .header("X-Service-Secret", &self.config.service_secret)
            .send()
            .await;
        tracing::info!("deregistered instance {id} from platform-api");
    }

    /// Initial registration with exponential backoff, retrying indefinitely
    /// until platform-api answers (mirrors Java sync + background retry).
    pub async fn register_with_backoff(&self) {
        let backoff_ms = [500u64, 1_000, 2_000, 4_000, 8_000, 15_000, 30_000];
        let mut i = 0usize;
        loop {
            match self.register_once().await {
                Ok(()) => {
                    tracing::info!(
                        "registered with platform-api as instance {} (tag: {})",
                        self.instance_id(),
                        if self.config.space_tag.is_empty() { "untagged" } else { &self.config.space_tag }
                    );
                    return;
                }
                Err(e) => {
                    let wait = backoff_ms[i.min(backoff_ms.len() - 1)];
                    tracing::warn!("platform-api registration failed: {e}; retry in {wait}ms");
                    tokio::time::sleep(Duration::from_millis(wait)).await;
                    i += 1;
                }
            }
        }
    }

    /// Spawn background lifecycle: NATS subscriptions + periodic re-register.
    /// Call after [`Registrar::register_with_backoff`]. Returns the NATS
    /// subscriptions (keep them alive for the process lifetime).
    pub async fn spawn_lifecycle(self: &Arc<Self>, nats: Option<NatsBus>) -> Vec<crate::nats::Subscription> {
        let mut subs = Vec::new();

        if let Some(bus) = nats {
            let me = Arc::clone(self);
            if let Ok(sub) = bus
                .subscribe("env.global.PLATFORM_RESTARTED", move |_msg| {
                    let me = Arc::clone(&me);
                    tokio::spawn(async move {
                        tokio::time::sleep(Duration::from_secs(2)).await;
                        // platform-api's registry version reset on restart;
                        // drop our monotonic baseline so the fresh snapshot is
                        // accepted instead of rejected as stale.
                        me.registry.reset_version();
                        if me.register_once().await.is_ok() {
                            tracing::info!("re-registered after PLATFORM_RESTARTED");
                        }
                    });
                })
                .await
            {
                subs.push(sub);
            }

            let me2 = Arc::clone(self);
            if let Ok(sub) = bus
                .subscribe("env.global.ENVIRONMENT_CHANGED", move |_msg| {
                    let me2 = Arc::clone(&me2);
                    tokio::spawn(async move {
                        me2.pull_snapshot().await;
                    });
                })
                .await
            {
                subs.push(sub);
            }
        }

        // Periodic re-register.
        let me = Arc::clone(self);
        tokio::spawn(async move {
            loop {
                tokio::time::sleep(RE_REGISTER_PERIOD).await;
                if me.register_once().await.is_ok() {
                    tracing::debug!("periodic re-register ok");
                }
            }
        });

        subs
    }
}
